ArduSub/motors.cpp

```diff
@@ -130,11 +130,111 @@ void Sub::motors_output()
 {
     // check if we are performing the motor test
     if (ap.motor_test) {
-        return; // Placeholder
+        verify_motor_test();
+    } else {
+        motors.set_interlock(true);
+        motors.output();
     }
-    motors.set_interlock(true);
-    motors.output();
 }
+static uint32_t last_do_motor_test_fail_ms = 0;
+static uint32_t last_do_motor_test_ms = 0;
+// Initialize new style motor test
+// Perform checks to see if it is ok to begin the motor test
+// Returns true if motor test has begun
+bool Sub::init_motor_test()
+{
+    uint32_t tnow = AP_HAL::millis();
+
+    // Ten second cooldown period required with no do_set_motor requests required
+    // after failure.
+    if (tnow < last_do_motor_test_fail_ms + 10000 && last_do_motor_test_fail_ms > 0) {
+        gcs().send_text(MAV_SEVERITY_CRITICAL, "10 second cool down required");
+    }
+
+    // check if safety switch has been pushed
+    if (hal.util->safety_switch_state() == AP_HAL::Util::SAFETY_DISARMED) {
+        gcs().send_text(MAV_SEVERITY_CRITICAL,"Disarm hardware safety switch before testing motors.");
+        return false;
+    }
+
+    // Make sure we are on the ground
+    if (!motors.armed()) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "Arm motors before testing motors.");
+        return false;
+    }
+
+    enable_motor_output(); // set all motor outputs to zero
+    ap.motor_test = true;
+
+    return true;
+}
+
+// Verify new style motor test
+// The motor test will fail if the interval between received
+// MAV_CMD_DO_SET_MOTOR requests exceeds a timeout period
+// Returns true if it is ok to proceed with new style motor test
+bool Sub::verify_motor_test()
+{
+    bool pass = true;
+
+    // Require at least 2 Hz incoming do_set_motor requests
+    if (AP_HAL::millis() > last_do_motor_test_ms + 500) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "Motor test timed out!");
+        pass = false;
+    }
+
+    if (!pass) {
+        ap.motor_test = false;
+        motors.armed(false); // disarm motors
+        last_do_motor_test_fail_ms = AP_HAL::millis();
+        return false;
+    }
+
+    return true;
+}
+
+bool Sub::handle_do_motor_test(mavlink_command_long_t command) {
+    last_do_motor_test_ms = AP_HAL::millis();
+
+    // If we are not already testing motors, initialize test
+    if(!ap.motor_test) {
+        if (!init_motor_test()) {
+            gcs().send_text(MAV_SEVERITY_WARNING, "motor test initialization failed!");
+            return false; // init fail
+        }
+    }
+
+    float motor_number = command.param1;
+    float throttle_type = command.param2;
+    float throttle = command.param3;
+    // float timeout_s = command.param4; // not used
```

Comment:
Are you going to use it in the future? It's much nicer than having to continuously send the message.

Reply:
No. I do not intend to use it. The idea is to spin at the speed that the user commands in real time.

![screenshot 2018-04-16 17 00 56](https://user-images.githubusercontent.com/8315108/39371751-8810e830-4a10-11e8-8a71-7c02f18b7026.png)

Reply:
Looks like a weird implementation to me, will depend on the MAVLink delay.

Reply:
![image](https://user-images.githubusercontent.com/8315108/39526943-3df71350-4dee-11e8-88e2-98a4290d335c.png)

Reply:
To me those UIs look just like the PILOT throttle type that you aren't implementing, the motor output changes according to what the user input is instead of a fixed value (which are the ones you are implementing).

Reply:
To me this is semantics. The spec is not clear. This UI is not the pilot's transmitter so I reject PILOT. What we are implementing works perfect and is clear to our users.

Reply:
>  It requires a fast repeated sending of a mavlink message to update the desired setpoints of the motors.

I understand that it is required in the way it is implemented, I just don't understand why you want that 😉

Anyway, I think I've exceeded my duty here, ultimately you are the vehicle maintainer and it's up to you on how things are implemented.